### PR TITLE
feat(command): 3rd party command query via messaging

### LINKED ISSUE #4078: Core command subscribes to query requests and publishes query responses to 3rd party requester via messaging
Core command to receive (subscribe) to 3rd party query requests via MQTT, fetch the required command metadata, package the metadata into a response message, and publish a query response message to a 3rd party query response topic.

### PR DESCRIPTION
**Pull request by hahattan**
allow core command to subscribe query request and publish
commands response to 3rd party requester with external MQTT

fix #4078

Signed-off-by: Chris Hung <chris@iotechsys.com>

**If your build fails**  due to your commit message not passing the build checks, please review the guidelines here: https://github.com/edgexfoundry/edgex-go/blob/main/.github/Contributing.md

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] I am not introducing a breaking change (if you are, flag in conventional commit message with `BREAKING CHANGE:` describing the break)
- [x] I am not introducing a new dependency (add notes below if you are)
- [x] I have added unit tests for the new feature or bug fix (if not, why?)
- [x] I have fully tested (add details below) this the new feature or bug fix (if not, why?)
- [ ] I have opened a PR for the related docs change (if not, why?)
  <link to docs PR>

## Testing Instructions

1. Run command service from this branch with `MESSAGEQUEUE_REQUIRED=true` environment variable
2. send query request message to `edgex/commandquery/request/<devicename>` topic.
For example send following message to `edgex/commandquery/request/Random-Boolean-Device`:
```json
{
  "correlationID": "14a42ea6-c394-41c3-8bcd-a29b9f5e6835",
  "apiVersion": "v2",
  "requestId": "e6e8a2f4-eb14-4649-9e2b-175247911369",
}
```
4. check the received message:
```json
{
  "ReceivedTopic": "edgex/commandquery/request/Random-Boolean-Device",
  "CorrelationID": "14a42ea6-c394-41c3-8bcd-a29b9f5e6835",
  "ApiVersion": "v2",
  "RequestID": "e6e8a2f4-eb14-4649-9e2b-175247911369",
  "ErrorCode": 0,
  "Payload": "eyJkZXZpY2VOYW1lIjoiUmFuZG9tLUJvb2xlYW4tRGV2aWNlIiwicHJvZmlsZU5hbWUiOiJSYW5kb20tQm9vbGVhbi1EZXZpY2UiLCJjb3JlQ29tbWFuZHMiOlt7Im5hbWUiOiJXcml0ZUJvb2xWYWx1ZSIsInNldCI6dHJ1ZSwicGF0aCI6Ii9hcGkvdjIvZGV2aWNlL25hbWUvUmFuZG9tLUJvb2xlYW4tRGV2aWNlL1dyaXRlQm9vbFZhbHVlIiwidXJsIjoiaHR0cDovL2VkZ2V4LWNvcmUtY29tbWFuZDo1OTg4MiIsInBhcmFtZXRlcnMiOlt7InJlc291cmNlTmFtZSI6IkJvb2wiLCJ2YWx1ZVR5cGUiOiJCb29sIn0seyJyZXNvdXJjZU5hbWUiOiJFbmFibGVSYW5kb21pemF0aW9uX0Jvb2wiLCJ2YWx1ZVR5cGUiOiJCb29sIn1dfSx7Im5hbWUiOiJXcml0ZUJvb2xBcnJheVZhbHVlIiwic2V0Ijp0cnVlLCJwYXRoIjoiL2FwaS92Mi9kZXZpY2UvbmFtZS9SYW5kb20tQm9vbGVhbi1EZXZpY2UvV3JpdGVCb29sQXJyYXlWYWx1ZSIsInVybCI6Imh0dHA6Ly9lZGdleC1jb3JlLWNvbW1hbmQ6NTk4ODIiLCJwYXJhbWV0ZXJzIjpbeyJyZXNvdXJjZU5hbWUiOiJCb29sQXJyYXkiLCJ2YWx1ZVR5cGUiOiJCb29sQXJyYXkifSx7InJlc291cmNlTmFtZSI6IkVuYWJsZVJhbmRvbWl6YXRpb25fQm9vbEFycmF5IiwidmFsdWVUeXBlIjoiQm9vbCJ9XX0seyJuYW1lIjoiQm9vbCIsImdldCI6dHJ1ZSwic2V0Ijp0cnVlLCJwYXRoIjoiL2FwaS92Mi9kZXZpY2UvbmFtZS9SYW5kb20tQm9vbGVhbi1EZXZpY2UvQm9vbCIsInVybCI6Imh0dHA6Ly9lZGdleC1jb3JlLWNvbW1hbmQ6NTk4ODIiLCJwYXJhbWV0ZXJzIjpbeyJyZXNvdXJjZU5hbWUiOiJCb29sIiwidmFsdWVUeXBlIjoiQm9vbCJ9XX0seyJuYW1lIjoiQm9vbEFycmF5IiwiZ2V0Ijp0cnVlLCJzZXQiOnRydWUsInBhdGgiOiIvYXBpL3YyL2RldmljZS9uYW1lL1JhbmRvbS1Cb29sZWFuLURldmljZS9Cb29sQXJyYXkiLCJ1cmwiOiJodHRwOi8vZWRnZXgtY29yZS1jb21tYW5kOjU5ODgyIiwicGFyYW1ldGVycyI6W3sicmVzb3VyY2VOYW1lIjoiQm9vbEFycmF5IiwidmFsdWVUeXBlIjoiQm9vbEFycmF5In1dfV19",
  "ContentType": "application/json",
  "QueryParams": null
}
```

### CODE REVIEW

File: internal/core/command/controller/messaging/external/external.go [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+		deviceName := topicLevels[len(topicLevels)-1]
+		if strings.EqualFold(deviceName, common.All) {
+			deviceName = common.All
+		}
+
+		var commands any
+		var edgexErr errors.EdgeX
+		switch deviceName {
+		case common.All:
+			offset, limit := 0, 20
```
Comment (lenny-intel) [inline internal/core/command/controller/messaging/external/external.go:74]: Are the the same default used for REST? They should be common constants.

File: internal/core/command/controller/messaging/external/external.go [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+			return
+		}
+
+		topicLevels := strings.Split(message.Topic(), "/")
+		deviceName := topicLevels[len(topicLevels)-1]
+		if strings.EqualFold(deviceName, common.All) {
+			deviceName = common.All
+		}
+
+		var commands any
```
Comment (lenny-intel) [inline internal/core/command/controller/messaging/external/external.go:70]: Love the use of `any` !

**Review by lenny-intel — CHANGES_REQUESTED** [review]
Missing unit tests. 
You will need to run mockery against `Client` at https://github.com/eclipse/paho.mqtt.golang/tree/v1.4.1
and then copy the generated file into new `mocks` folder under the `messaging` folder.

File: internal/core/command/controller/messaging/external/external_test.go [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+			client.On("Publish", testResponseTopic, byte(0), true, mock.Anything).Return(token)
+
+			fn := commandQueryHandler(testResponseTopic, 0, true, dic)
+			fn(client, message)
+			lc.AssertCalled(t, "Debugf", mock.Anything, mock.Anything, mock.Anything)
+			if tt.expectedError {
+				lc.AssertCalled(t, "Error", mock.Anything)
+			}
+		})
+	}
```
Comment (lenny-intel) [inline internal/core/command/controller/messaging/external/external_test.go:149]: Nice Job!

File: internal/core/command/controller/messaging/external/external.go [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+	"github.com/edgexfoundry/edgex-go/internal/core/command/application"
+	"github.com/edgexfoundry/edgex-go/internal/core/command/container"
+)
+
+const (
+	RequestQueryTopic  = "RequestQueryTopic"
+	ResponseQueryTopic = "ResponseQueryTopic"
+)
+
+func OnConnectHandler(dic *di.Container) mqtt.OnConnectHandler {
```
Comment (lenny-intel) [inline internal/core/command/controller/messaging/external/external.go:32]: Can now also unit test this function

File: internal/core/command/controller/messaging/external/external.go [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+		lc := bootstrapContainer.LoggingClientFrom(dic.Get)
+		config := container.ConfigurationFrom(dic.Get)
+		externalTopics := config.MessageQueue.External.Topics
+		qos := config.MessageQueue.External.QoS
+		retain := config.MessageQueue.External.Retain
+
+		requestQueryTopic := externalTopics[RequestQueryTopic]
+		responseQueryTopic := externalTopics[ResponseQueryTopic]
+		if token := client.Subscribe(requestQueryTopic, qos, commandQueryHandler(responseQueryTopic, qos, retain, dic)); token.Wait() && token.Error() != nil {
+			client.Disconnect(0)
```
Comment (lenny-intel) [inline internal/core/command/controller/messaging/external/external.go:43]: Should we be disconnecting here?

**Review by lenny-intel — CHANGES_REQUESTED** [review]

File: internal/core/command/controller/messaging/external/external.go (hunk 5 above)
Comment (hahattan) [reply internal/core/command/controller/messaging/external/external.go:43]: I followed app-sdk's behavior:
https://github.com/edgexfoundry/app-functions-sdk-go/blob/main/internal/trigger/mqtt/mqtt.go#L157-L164

But considering we have command request topic to subscribe as well should we just log the error and not disconnect here ?

File: internal/core/command/controller/messaging/external/external.go [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+		lc := bootstrapContainer.LoggingClientFrom(dic.Get)
+
+		requestEnvelope, err := types.NewMessageEnvelopeFromJSON(message.Payload())
+		if err != nil {
+			responseEnvelope = types.NewMessageEnvelopeWithError(requestEnvelope.RequestID, err.Error())
+			publishMessage(client, responseTopic, qos, retain, responseEnvelope, lc)
+			return
+		}
+
+		// example topic scheme: edgex/commandquery/request/#
```
Comment (lenny-intel) [inline internal/core/command/controller/messaging/external/external.go:63]: Actually the received topic doesn't have the wild card `#`, 
```suggestion
		// example topic scheme: edgex/commandquery/request/<device>
```

File: internal/core/command/controller/messaging/external/external.go (hunk 5 above)
Comment (lenny-intel) [reply internal/core/command/controller/messaging/external/external.go]: K, IMO we should not disconnect here. We have the disconnect when exiting which should suffice. 
I will file an issue on the SDK to do the same.

**Review by lenny-intel — CHANGES_REQUESTED** [review]

**Review by lenny-intel — APPROVED** [review]
LGTM

**Pull request merged**